src/ILCompiler.ReadyToRun/src/ObjectWriter/R2RPEBuilder.cs

```diff
@@ -865,28 +865,60 @@ public static PEHeaderBuilder Copy(PEHeaders peHeaders, TargetDetails target)
                 imageCharacteristics |= Characteristics.LargeAddressAware;
             }
 
+            int fileAlignment = 0x200;
+            if (!target.IsWindows && !is64BitTarget)
+            {
+                // To minimize wasted VA space on 32 bit systems align file to page bounaries (presumed to be 4K).
+                fileAlignment = 0x1000;
+            }
+
+            DllCharacteristics dllCharacteristics = 0;
+
+            if (!is64BitTarget)
+            {
+                dllCharacteristics |= DllCharacteristics.NoSeh;
+            }
+
+            if (target.Architecture == TargetArchitecture.ARM)
+            {
+                // Images without NX compat bit set fail to load on ARM
+                dllCharacteristics = DllCharacteristics.NxCompatible;
+            }
+
+            // Copy over selected DLL characteristics bits from IL image
+            dllCharacteristics |= peHeaders.PEHeader.DllCharacteristics &
+                (DllCharacteristics.NxCompatible | DllCharacteristics.TerminalServerAware | DllCharacteristics.AppContainer);
+
+            dllCharacteristics |= DllCharacteristics.DynamicBase;
+
+            if (is64BitTarget)
+            {
+                dllCharacteristics |= DllCharacteristics.HighEntropyVirtualAddressSpace;
+            }
+
             return new PEHeaderBuilder(
                 machine: target.MachineFromTarget(),
                 // On Linux, we must match the bottom 12 bits of section RVA's to their file offsets. For this reason
                 // we need the same alignment for both.
-                sectionAlignment: (target.IsWindows ? peHeaders.PEHeader.SectionAlignment : peHeaders.PEHeader.FileAlignment),
-                fileAlignment: peHeaders.PEHeader.FileAlignment,
+                sectionAlignment: (target.IsWindows || !is64BitTarget ? 0x1000 : fileAlignment),
```

Comment:
Nit - it's not your fault but it might make the code a tad easier to read if we also calculated the sectionAligment value upfront like you're newly doing for fileAlignment, especially as the condition has become more complicated.